### ISSUE TITLE
Does chdb supports orm? If supports, how to use?

### ISSUE THREAD

**Issue by flyly0755** [chdb-io/chdb#213]
Use orm, we doesn't need to write raw sql to do crud operation, instead of using sqlalchemy package to deal with database.
Usually, with one orm class to map one database table. for example as below:
```python
from sqlalchemy import Column
from clickhouse_sqlalchemy import engines
from clickhouse_sqlalchemy.ext.declarative import declarative_base
from clickhouse_sqlalchemy.types import String, UInt32

ChBase = declarative_base()

class FileInfo(ChBase):
    __tablename__ = 'FileInfo'
    FILE_ID = Column(UInt32, primary_key=True)
    filename = Column(String)
    filepath = Column(String)
    filemd5 = Column(String(32))
    __table_args__ = (
        engines.MergeTree(order_by=('FILE_ID',),
                          primary_key=('FILE_ID',)),
        {'comment': 'FileInfo Table in clickhouse'}
    )

from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

ckuser = 'ckuser'
ckpwd = 'ckpwd'
ckip = 'ckip'
ckport = '8123'
ckdbname = 'testdb'
uri = f"clickhouse://{ckuser}:{ckpwd}@{ckip}:{ckport}/{ckdbname}"
engine = create_engine(uri, echo=False)
session = sessionmaker(bind=engine)()
session.execute('SELECT 1')
FileInfo.__table__.create(engine)  # create table FileInfo

session.query(FileInfo).filter_by(FILE_ID=1).all()  # equal to sql: select * from FileInfo where FILE_ID=1
```
chdb is a superb package, very useful for testing and debug👍👍👍
But all code related with clickhouse uses orm in my softwore project, so I want to whether chdb supports orm?
uri is a very important parameter.
If supports, how to set uri? something like "clickhouse+**chdb**://xxx"?

**Comment by auxten** [comment]
chDB do have Python DB-API support. See https://github.com/chdb-io/chdb/blob/main/examples/dbapi.py
I didn't try it with SQLAlchemy. You can give it a try. If any problem, please tell me here or on discord.

**Comment by flyly0755** [comment]
seems not support, coz uri is a fundamental parameter. with uri(user:password@dbip:port/dbname), client can connect with database.

**Comment by flyly0755** [comment]
@auxten I also ask sqlalchemy for this question ^_^
Need extra database dialect code development to support this function.
https://github.com/sqlalchemy/sqlalchemy/discussions/11319
chdb and sqlalchemy both sides need to negotiate this work.

**Comment by auxten** [comment]
Thank you! I will look into this tomorrow. BTW, are you interested to contribute this for chDB?😉

**Comment by flyly0755** [comment]
Yes, I am interested😊, but I am not familar with lower level logics of sqlalchemy dialect.
Maybe in the future after I understand these base knowledge, I can make a help.
By the way, I have found the link of supported sqlalchemy dialect
https://docs.sqlalchemy.org/en/14/dialects/

**Comment by rominf** [comment]
I spent a few hours to prepare POC: https://github.com/rominf/clickhouse-sqlalchemy/tree/rominf-chdb. It is still on early stages (not production ready), but kinda works for poking. I had to make some workarounds for known issues: absence of persistence between calls for Memory engine and default database: https://github.com/chdb-io/chdb/issues/262#issuecomment-2325570573 and https://github.com/chdb-io/chdb/issues/225#issuecomment-2154217779. Also, the conversion for complex types does not work (this contributes to most test failures). Currently, 25/177 tests of testing suite testing chDB driver FAIL (in other words, 152/177 PASS).

To play with it, install it as usual from the branch above and then write something like:
```
from sqlalchemy import create_engine, Column, MetaData

from clickhouse_sqlalchemy import (
    Table, make_session, get_declarative_base, types, engines
)

uri = 'clickhouse+chdb:///test?path=/tmp/chdb-test'

engine = create_engine(uri)
session = make_session(engine)
metadata = MetaData()

Base = get_declarative_base(metadata=metadata)

class Rate(Base):
    day = Column(types.Date, primary_key=True)
    value = Column(types.Int32)
    __table_args__ = (
        engines.Log(),
    )

metadata.create_all(bind=engine)
from datetime import date, timedelta

from sqlalchemy import func

today = date.today()
rates = [
    {'day': today - timedelta(i), 'value': 200 - i}
    for i in range(100)
]
session.execute(Rate.__table__.insert(), rates)
session.query(Rate).count()  # Outputs 100
```

Feel free to borrow my code/ideas/contact me for working together/etc. to get this driver fully working!

**Comment by zbs** [comment]
Hi @rominf. Thanks for taking a stab at the chdb driver. One issue I observed with your code is that the connection doesn't seem to close when you exit the with block
```
with engine.connect() as conn:
    conn.execute(text('create view bar as select 1'))
chdb.connect('foo')
````
The last line fails with
```
RuntimeError: Another connection is already active with different path. Old path = :memory:, new path = /path/to/cwd/foo, please close the existing connection first.
````

**Comment by wudidapaopao** [comment]
Hi @zbs, before creating a new `Connection` or `Session`, you need to manually call the `close()` function to close the previous `Connection` or `Session`. However, Session can be automatically closed using the `with .. as ..` syntax. Here are some examples.
```
import chdb
from chdb import session
test_db_path = "test_db_path"

with session.Session(test_db_path) as sess:
    sess.query("SELECT 1").show()

sess = session.Session(test_db_path)
sess.query("SELECT 1").show()
sess.close()

conn = chdb.connect(test_db_path)
conn.query("SELECT 1").show()
conn.close()

conn = chdb.connect(test_db_path)
conn.query("SELECT 1").show()
conn.close()
```

**Comment by zbs** [comment]
Right but for SQLAlchemy core usage, namely using 
```
with engine.connect() as conn:
   # do something with connection
```
It doesn't close the connection given the above example. Shouldn't need to explicitly call close with `contextmanager` / `with` block, especially if `Session` is capable of closing it this way

**Comment by wudidapaopao** [comment]
Yes, currently, `Connection` cannot be used with the `with ... as ...` syntax. You can use `Session` as an alternative. We may also consider enabling `Connection` to support the `with ... as ...` syntax in the future, which would only require very minor modifications.

**Comment by zbs** [comment]
What modifications would that require? 

Also, is there an effort to turn the prototype chdb driver code [above](https://github.com/chdb-io/chdb/issues/213#issuecomment-2351008782) into a property python package (or a component of `chdb`)?

**Comment by wudidapaopao** [comment]
I'm sorry, but I misunderstood your question earlier. I thought you were referring to chdb's Connection.

Are you currently using the code from https://github.com/rominf/clickhouse-sqlalchemy/tree/rominf-chdb? Regarding the question of whether chdb should be adapted to the [above](https://github.com/chdb-io/chdb/issues/213#issuecomment-2351008782) as soon as possible, we will discuss it later. If you have any ideas, you are more than welcome to contribute.

**Comment by zbs** [comment]
Yes exactly: I was referring to the SQLAlchemy connection, not the chdb connection. Even if you have a patch for his chdb driver to support proper connection closing, I would appreciate having it ASAP so I can use the driver until an official version is released.

**Comment by wudidapaopao** [comment]
It is relatively easy to make the chdb Connection object implement automatic closing. However, if it is necessary for his chdb driver to also support automatic closing, let me first take a look at his code.

**Comment by wudidapaopao** [comment]
Hi @zbs, I've submitted a [PR](https://github.com/rominf/clickhouse-sqlalchemy/pull/1) to the `rominf/clickhouse-sqlalchemy` repository, which adds support for automatically closing chdb connections. Feel free to give it a try if you need it.

**Comment by rominf** [comment]
Sorry for the absence of replies. I merged @wudidapaopao PR, please give it a try.

**Comment by zbs** [comment]
Thank you. Will try and report back

**Comment by zbs** [comment]
@wudidapaopao that did the trick! Thank you

**Comment by ShawnChen-Sirius** [comment]
Resolved — the reporter confirmed the SQLAlchemy/ORM approach works. Closing.
